File: frontend/admin/js/content.js
/* ═══════════════ CONTENT ═══════════════
   The three things a shop edits that aren't products or orders: which
   designs are offered as templates, which reviews are shown, and the print
   zones the studio measures against. */

const contState = {scope:'all'};

async function renderContent(el, sub){
  const tabs = [['design-templates','Design templates'],
                ['review-moderation','Review moderation'],
                ['product-photos','Product photos']];
  el.innerHTML = `<div class="row" style="margin-bottom:18px">
    ${tabs.map(([k,label])=>`<button class="btn btn-sm ${(sub||'design-templates')===k?'btn-primary':'btn-quiet'}"
      onclick="goTo('content','${k}')">${esc(label)}</button>`).join('')}
  </div><div id="contBody"></div>`;
  const body = document.getElementById('contBody');
  if(sub === 'review-moderation') return renderModeration(body);
  if(sub === 'product-photos')    return renderPhotos(body);
  return renderTemplates(body);
}

/* ── Templates ───────────────────────────────────────────────── */
async function renderTemplates(el){
  el.innerHTML = '<div class="empty">Loading…</div>';
  const d = await api('/api/admin/content/templates');
  if(!d.ok){ el.innerHTML = `<div class="empty">${esc(d.error)}</div>`; return; }
  /* The thumbnail is a button, not decoration — it's the obvious thing to
     click when you want a better look, and it was doing nothing. */
  const card = (t, isTemplate) => `<div class="tpl">
    <button class="tpl-thumb" onclick="viewDesign(${t.id})"
      title="Preview ${esc(t.name)}">${t.thumb?`<img src="${esc(t.thumb)}" alt="">`
      :'<span class="material-symbols-outlined">image</span>'}</button>
    <div class="tpl-meta">
      <b>${esc(t.name)}</b>
      <span class="tiny dim">${esc(t.product)} · ${esc(t.author||'unknown')} · ${esc(fmtDate(t.created))}</span>
    </div>
    <div class="row" style="gap:6px">
      <button class="btn btn-quiet btn-sm" onclick="viewDesign(${t.id})">View</button>
      ${isTemplate ? `
        <input type="text" value="${t.sort}" style="width:64px"
          onchange="setTemplate(${t.id},{sort:this.value})" aria-label="Sort order">
        <button class="btn btn-quiet btn-sm" onclick="setTemplate(${t.id},{is_template:false})">
          Unpublish</button>`
        : `<button class="btn btn-primary btn-sm" onclick="setTemplate(${t.id},{is_template:true})">
          Publish</button>`}
    </div>
  </div>`;
  el.innerHTML = `
    <div class="card">
      <h2>Published templates (${d.templates.length})</h2>
      <p class="tiny muted">Shown in the studio's "Start from template" picker,
        in the order below. Unpublishing takes it out of the picker — the
        design itself stays in its owner's account.</p>
      ${d.templates.map(t=>card(t,true)).join('') || '<div class="empty">None published.</div>'}
    </div>
    <div class="card">
      <h2>Recent customer designs</h2>
      <p class="tiny muted">A good template is usually something a customer
        already made. Publishing copies nothing — it flags this design, so
        the customer's own copy is what everyone starts from.</p>
      ${d.candidates.map(t=>card(t,false)).join('') || '<div class="empty">No saved designs yet.</div>'}
    </div>`;
}
async function setTemplate(id, patch){
  const d = await api('/api/admin/content/templates/'+id, patch);
  if(!d.ok){ toast(d.error); return; }
  toast('Updated');
  renderTemplates(document.getElementById('contBody'));
  if(document.getElementById('dlgDesign')) viewDesign(id);   // keep it in step
}

/* ── Design preview ──────────────────────────────────────────── */
/* The thumbnail is what the studio actually rendered, so it is the design.
   What it can't tell you is what the text says at readable size, or whether
   any layer is a customer's uploaded logo — publishing one of those hands
   it to every visitor. Both are spelled out here. */
async function viewDesign(id){
  let box = document.getElementById('dlgDesign');
  if(!box){
    box = document.createElement('div');
    box.id = 'dlgDesign';
    box.className = 'dlg';
    box.onclick = e => { if(e.target === box) closeDesign(); };
    document.body.appendChild(box);
    addEventListener('keydown', escDesign);
  }
  box.innerHTML = '<div class="dlg-card"><div class="empty">Loading…</div></div>';
  const d = await api('/api/admin/content/templates/'+id);
  if(!d.ok){ box.innerHTML = `<div class="dlg-card"><div class="empty">${esc(d.error)}</div></div>`; return; }
  const g = d.design;
  const layerList = side => (g.layers[side]||[]).map(l => l.kind==='text'
    ? `<li><b>“${esc(l.text)}”</b><span class="tiny dim"> ${esc(l.font)}
         · ${Math.round(l.size||0)}px <span class="swatch-dot"
         style="background:${esc(l.color||'#000')}"></span>${esc(l.color||'')}</span></li>`
    : `<li><b>Uploaded image</b><span class="tiny dim"> ${l.w}×${l.h} px on canvas</span></li>`
  ).join('') || '<li class="tiny dim">Nothing on this side.</li>';

  box.innerHTML = `<div class="dlg-card" role="dialog" aria-modal="true" aria-label="Design preview">
    <div class="drawer-head">
      <div><h2>${esc(g.name)}</h2>
        <p class="tiny muted">${esc(g.product)} · ${esc(g.author||'unknown')}
          ${g.author_email?'· '+esc(g.author_email):''} · ${esc(fmtDate(g.created))}</p></div>
      <button class="icon-btn" onclick="closeDesign()" aria-label="Close">
        <span class="material-symbols-outlined">close</span></button>
    </div>
    <div class="dlg-body">
      ${g.uploaded_images ? `<div class="alert warn">
        <span class="material-symbols-outlined">warning</span>
        <span><b>Contains ${g.uploaded_images} uploaded image(s).</b> If that's the
          customer's own logo or artwork, publishing it as a template hands it to
          every visitor. Check before you do.</span>
      </div>` : ''}
      <div class="dlg-preview" style="--swatch:${esc(g.shirt_color)}">
        ${g.thumb ? `<img src="${esc(g.thumb)}" alt="Preview of ${esc(g.name)}">`
          : '<div class="empty">No preview was saved with this design.</div>'}
      </div>
      <p class="tiny dim" style="text-align:center">Garment colour
        <span class="swatch-dot" style="background:${esc(g.shirt_color)}"></span>
        ${esc(g.shirt_color)} · rendered by the studio when it was saved</p>

      <div class="grid g2" style="margin-top:16px">
        <div><h3 class="tiny muted lbl">Front</h3><ul class="layer-list">${layerList('front')}</ul></div>
        <div><h3 class="tiny muted lbl">Back</h3><ul class="layer-list">${layerList('back')}</ul></div>
      </div>

      <div class="row" style="margin-top:18px">
        ${g.is_template
          ? `<a class="btn btn-quiet" href="/#/studio" target="_blank" rel="noopener"
               onclick="toast('Open it from the studio\\'s template picker')">Open the storefront</a>
             <button class="btn btn-quiet" onclick="setTemplate(${g.id},{is_template:false})">
               Unpublish</button>`
          : `<button class="btn btn-primary" onclick="setTemplate(${g.id},{is_template:true})">
               Publish as template</button>`}
      </div>
      ${!g.is_template ? `<p class="tiny dim" style="margin-top:8px">
        Unpublished designs stay private to their owner — this preview is the
        only way to see one, and opening it in the studio would mean loading
        someone else's saved work.</p>` : ''}
    </div>
  </div>`;
}
function closeDesign(){
  const box = document.getElementById('dlgDesign');
  if(box) box.remove();
  removeEventListener('keydown', escDesign);
}
function escDesign(e){ if(e.key === 'Escape') closeDesign(); }

/* ── Review moderation ───────────────────────────────────────── */
async function renderModeration(el){
  el.innerHTML = '<div class="empty">Loading…</div>';
  const d = await api('/api/admin/content/reviews?scope='+contState.scope);
  if(!d.ok){ el.innerHTML = `<div class="empty">${esc(d.error)}</div>`; return; }
  const f = [['all','All',d.counts.all],['visible','Visible',d.counts.visible],
             ['low','1–2 star',d.counts.low],['hidden','Hidden',d.counts.hidden]];
  el.innerHTML = `
    <div class="row" style="margin-bottom:14px">
      ${f.map(([k,label,n])=>`<button class="chip ${contState.scope===k?'on':''}"
        onclick="setReviewScope('${k}')">${esc(label)} <b>${n}</b></button>`).join('')}
    </div>
    ${d.reviews.map(r=>`<div class="card ${r.hidden?'row-off':''}">
      <div class="spread" style="margin-bottom:8px">
        <div>
          <b>${'★'.repeat(r.rating)}${'☆'.repeat(5-r.rating)}</b>
          <span class="tiny dim"> ${esc(r.product)} · ${esc(r.author||'—')}
            · ${esc(fmtDate(r.created))}</span>
        </div>
        ${r.hidden ? `<button class="btn btn-quiet btn-sm" onclick="moderate(${r.id},false)">
            Restore</button>`
          : `<button class="btn btn-quiet btn-sm" onclick="moderate(${r.id},true)">Hide</button>`}
      </div>
      <p style="font-size:13.5px;line-height:1.55">${esc(r.body)}</p>
      ${r.hidden?`<p class="tiny" style="color:var(--pink-ink);margin-top:8px">
        Hidden — ${esc(r.hidden_reason)}</p>`:''}
    </div>`).join('') || '<div class="empty">Nothing here.</div>'}
    <p class="tiny dim">Hiding takes a review off the product page and out of
      its rating average. Nothing is deleted — the row stays, along with who
      hid it and why.</p>`;
}
function setReviewScope(s){ contState.scope=s; renderModeration(document.getElementById('contBody')); }
async function moderate(id, hide){
  let reason = '';
  if(hide){
    reason = prompt('Why is this being hidden? (kept on the record)') || '';
    if(!reason.trim()) return;
  }
  const d = await api('/api/admin/content/reviews/'+id, {hidden:hide, reason});
  if(!d.ok){ toast(d.error); return; }
  toast(hide?'Hidden':'Restored');
  renderModeration(document.getElementById('contBody'));
}

/* ── Photos + print zones ────────────────────────────────────── */
async function renderPhotos(el){
  el.innerHTML = '<div class="empty">Loading…</div>';
  const d = await api('/api/admin/content/photos');
  if(!d.ok){ el.innerHTML = `<div class="empty">${esc(d.error)}</div>`; return; }
  el.innerHTML = `
    <div class="alert warn" style="margin-bottom:16px">
      <span class="material-symbols-outlined">straighten</span>
      <span><b>These measurements drive every garment size.</b> The studio
        scales the zone up and down from this reference by the size chart, so
        one wrong figure here is wrong on all six sizes at once. Measure a real
        blank and the maximum your press can lay down — don't guess.</span>
    </div>
    ${d.photos.map(p=>`<div class="card">
      <div class="spread" style="margin-bottom:10px">
        <div><h2>${esc(p.product)} — ${esc(p.side)}</h2>
          <p class="tiny muted" style="font-family:monospace">${esc(p.file)}</p></div>
        <img class="photo-thumb" src="/${esc(p.file)}" alt=""
          onerror="this.classList.add('gone');this.alt='missing'">
      </div>
      ${p.zone ? `
        <div class="form-grid zone-grid">
          ${[['cx','Centre X (px)'],['cy','Centre Y (px)'],
             ['w','Width (px)'],['h','Height (px)'],
             ['cm_w','Width (cm)'],['cm_h','Height (cm)']].map(([f,label])=>`
            <label class="field"><span>${esc(label)}</span>
              <input id="z_${esc(p.key)}_${f}" type="text"
                value="${f==='cm_w'?p.zone.cmW:f==='cm_h'?p.zone.cmH:p.zone[f]}"></label>`).join('')}
        </div>
        <button class="btn btn-primary btn-sm" onclick="saveZone('${esc(p.key)}')">Save zone</button>
        <p class="tiny dim" style="margin-top:8px">Pixels are measured on the
          mockup photo at its native resolution; centimetres are the real printable size on
          that garment.</p>`
        : `<div class="alert warn"><span class="material-symbols-outlined">error</span>
           <span>No zone recorded — the studio falls back to the round-neck tee's
             measurements, so this mockup prints at the wrong dimensions.</span></div>`}
    </div>`).join('')}
    ${d.no_photo.length?`<div class="card">
      <h2>Drawn without a photo</h2>
      <p class="tiny muted">${d.no_photo.map(esc).join(', ')} — no mockup file, so
        the studio draws a vector silhouette and uses the fallback print area in
        studio.js. That's by design, not a fault; add a photo to change it.</p>
    </div>`:''}
    ${d.orphan_zones.length?`<div class="card">
      <h2>Zones with no photo</h2>
      <p class="tiny muted">${d.orphan_zones.map(esc).join(', ')} — measurements
        for a mockup that isn't on disk. Harmless, but nothing reads them.</p>
    </div>`:''}
    <p class="tiny dim">Photos ship with the code and aren't uploadable here:
      everything outside the data volume is replaced on each deploy, so an
      uploaded file would disappear at the next push with nothing to connect
      the two. Replacing a mockup is a commit.</p>`;
}
async function saveZone(key){
  const num = f => document.getElementById('z_'+key+'_'+f).value;
  const d = await api('/api/admin/content/zones/'+encodeURIComponent(key), {
    cx:num('cx'), cy:num('cy'), w:num('w'), h:num('h'),
    cm_w:num('cm_w'), cm_h:num('cm_h'),
  });
  if(!d.ok){ toast(d.error); return; }
  toast('Saved — live in the studio on the next page load');
}

File: backend/content.py
# ── Content blueprint: templates, review moderation, print zones ──
#
# The three things a shop edits that aren't products or orders.
#
# Review moderation hides, never deletes: a review taken down for language
# is still evidence the purchase happened, and "who removed this and why"
# is a question that gets asked.
#
# Print zones used to live in js/mockup-data.js, which meant correcting a
# measurement was a code change. They matter more than they look: every
# garment size is derived from the reference zone by scaling, so one wrong
# figure is wrong on all six sizes at once.
import json

from flask import Blueprint, request, jsonify, session

from db import get_db
from permissions import require

content_bp = Blueprint("content", __name__, url_prefix="/api/admin/content")


def zones_payload(db=None):
    """{mock_key: {cx,cy,w,h,cmW,cmH}} in the shape mockup-data.js uses, so
    the storefront can merge it straight over the file defaults."""
    db = db or get_db()
    rows = db.execute("SELECT * FROM print_zones").fetchall()
    return {r["mock_key"]: {"cx": r["cx"], "cy": r["cy"], "w": r["w"], "h": r["h"],
                            "cmW": r["cm_w"], "cmH": r["cm_h"]} for r in rows}


# ── Design templates ─────────────────────────────────────────────
@content_bp.route("/templates")
@require("content")
def list_templates():
    db = get_db()
    rows = db.execute(
        """SELECT d.id,d.name,d.product_id,d.shirt_color,d.thumb,d.is_template,
                  d.sort,d.created,u.name AS author
           FROM saved_designs d LEFT JOIN users u ON u.id=d.user_id
           WHERE d.is_template=1 ORDER BY d.sort, d.id""").fetchall()
    # Recent customer designs, offered as promotion candidates. A good
    # template is usually something a customer already made, not something
    # invented in an empty studio.
    cand = db.execute(
        """SELECT d.id,d.name,d.product_id,d.thumb,d.created,u.name AS author
           FROM saved_designs d LEFT JOIN users u ON u.id=d.user_id
           WHERE d.is_template=0 ORDER BY d.id DESC LIMIT 24""").fetchall()
    pub = lambda r, extra=(): dict(
        {"id": r["id"], "name": r["name"], "product": r["product_id"],
         "thumb": r["thumb"], "created": r["created"], "author": r["author"]},
        **{k: r[k] for k in extra})
    return jsonify(ok=True,
                   templates=[pub(r, ("sort",)) for r in rows],
                   candidates=[pub(r) for r in cand])


@content_bp.route("/templates/<int:did>")
@require("content")
def template_detail(did):
    """Enough to decide whether to publish it. A 56px thumbnail isn't —
    publishing puts someone's design in front of every visitor, and the two
    things that decide it are what the artwork actually says and whether any
    of it is a customer's own uploaded logo."""
    db = get_db()
    r = db.execute(
        """SELECT d.*, u.name AS author, u.email AS author_email
           FROM saved_designs d LEFT JOIN users u ON u.id=d.user_id
           WHERE d.id=?""", (did,)).fetchone()
    if not r:
        return jsonify(ok=False, error="No such design."), 404
    try:
        layers = json.loads(r["layers_json"]) or {}
    except (ValueError, TypeError):
        layers = {}

    sides, uploaded = {}, 0
    for side in ("front", "back"):
        out = []
        for l in layers.get(side) or []:
            if not isinstance(l, dict):
                continue
            if l.get("type") == "text":
                out.append({"kind": "text", "text": l.get("text") or "",
                            "font": l.get("font") or "", "color": l.get("color") or "",
                            "size": l.get("size")})
            else:
                uploaded += 1
                out.append({"kind": "image",
                            "w": round(l.get("w") or 0), "h": round(l.get("h") or 0)})
        sides[side] = out

    prod = db.execute("SELECT name FROM products WHERE slug=?", (r["product_id"],)).fetchone()
    return jsonify(ok=True, design={
        "id": r["id"], "name": r["name"], "product_id": r["product_id"],
        "product": prod["name"] if prod else r["product_id"],
        "shirt_color": r["shirt_color"], "thumb": r["thumb"],
        "is_template": bool(r["is_template"]), "created": r["created"],
        "author": r["author"], "author_email": r["author_email"],
        "layers": sides,
        # Publishing a design built on a customer's uploaded logo hands that
        # logo to everyone. The panel can't tell whose it is — it can only
        # make sure nobody publishes one without noticing.
        "uploaded_images": uploaded,
    })


@content_bp.route("/templates/<int:did>", methods=["POST"])
@require("content")
def set_template(did):
    from admin_api import log
    d = request.get_json(force=True, silent=True) or {}
    db = get_db()
    row = db.execute("SELECT * FROM saved_designs WHERE id=?", (did,)).fetchone()
    if not row:
        return jsonify(ok=False, error="No such design."), 404
    if "is_template" in d:
        on = 1 if d["is_template"] else 0
        db.execute("UPDATE saved_designs SET is_template=? WHERE id=?", (on, did))
        log(db, "design", did, "template" if on else "untemplate", None, None, row["name"])
    if "name" in d:
        name = (d["name"] or "").strip()[:80]
        if name:
            db.execute("UPDATE saved_designs SET name=? WHERE id=?", (name, did))
    if "sort" in d:
        try:
            db.execute("UPDATE saved_designs SET sort=? WHERE id=?", (int(d["sort"]), did))
        except (TypeError, ValueError):
            pass
    db.commit()
    return jsonify(ok=True)


# ── Review moderation ────────────────────────────────────────────
@content_bp.route("/reviews")
@require("content")
def list_reviews_admin():
    db = get_db()
    scope = (request.args.get("scope") or "all").strip()
    sql = """SELECT r.*, u.name AS author, u.email AS author_email, p.name AS product_name
             FROM reviews r
             LEFT JOIN users u ON u.id=r.user_id
             LEFT JOIN products p ON p.slug=r.product_id
             WHERE 1=1"""
    if scope == "hidden":
        sql += " AND r.hidden=1"
    elif scope == "visible":
        sql += " AND r.hidden=0"
    elif scope == "low":
        sql += " AND r.rating<=2 AND r.hidden=0"
    sql += " ORDER BY r.id DESC LIMIT 300"
    rows = db.execute(sql).fetchall()
    counts = {
        "all": db.execute("SELECT COUNT(*) c FROM reviews").fetchone()["c"],
        "hidden": db.execute("SELECT COUNT(*) c FROM reviews WHERE hidden=1").fetchone()["c"],
        "low": db.execute(
            "SELECT COUNT(*) c FROM reviews WHERE rating<=2 AND hidden=0").fetchone()["c"],
    }
    counts["visible"] = counts["all"] - counts["hidden"]
    return jsonify(ok=True, scope=scope, counts=counts, reviews=[{
        "id": r["id"], "product": r["product_name"] or r["product_id"],
        "product_id": r["product_id"], "rating": r["rating"], "body": r["body"],
        "author": r["author"], "author_email": r["author_email"],
        "created": r["created"], "hidden": bool(r["hidden"]),
        "hidden_reason": r["hidden_reason"],
    } for r in rows])


@content_bp.route("/reviews/<int:rid>", methods=["POST"])
@require("content")
def moderate_review(rid):
    from admin_api import log
    d = request.get_json(force=True, silent=True) or {}
    hide = 1 if d.get("hidden") else 0
    reason = (d.get("reason") or "").strip()[:200]
    if hide and len(reason) < 3:
        return jsonify(ok=False, error="Say why it's being hidden."), 400
    db = get_db()
    r = db.execute("SELECT * FROM reviews WHERE id=?", (rid,)).fetchone()
    if not r:
        return jsonify(ok=False, error="No such review."), 404
    db.execute("UPDATE reviews SET hidden=?, hidden_reason=? WHERE id=?",
               (hide, reason if hide else "", rid))
    log(db, "review", rid, "hidden" if hide else "restored", None, None, reason)
    db.commit()
    return jsonify(ok=True)


# ── Product photos + print zones ─────────────────────────────────
@content_bp.route("/photos")
@require("content")
def photos():
    """What each product's mockup is, and the zone measured on it.

    The image files themselves ship with the code — they are not uploadable
    here on purpose. Railway's filesystem outside the volume is replaced on
    every deploy, so an uploaded photo would vanish at the next push and
    nobody would connect the two events. Replacing a photo is a commit.
    """
    import os
    db = get_db()
    zones = zones_payload(db)
    names = {p["slug"]: p["name"]
             for p in db.execute("SELECT slug,name FROM products").fetchall()}

    # Built from the files that are actually on disk, not from the product
    # list. A product with no mockup (the tote) draws a vector silhouette
    # instead, so inventing a row for it would report a problem that isn't
    # one — and inventing `tb_back` would report a second.
    mdir = os.path.join(os.path.dirname(__file__), "..", "frontend", "mockups")
    try:
        files = sorted(f for f in os.listdir(mdir) if f.lower().endswith(".jpg"))
    except OSError:
        files = []

    out = []
    for f in files:
        key = os.path.splitext(f)[0]
        slug = key[:-5] if key.endswith("_back") else key
        try:
            size = os.path.getsize(os.path.join(mdir, f))
        except OSError:
            size = 0
        out.append({
            "product": names.get(slug, slug), "product_id": slug,
            "side": "back" if key.endswith("_back") else "front",
            "key": key, "file": "mockups/%s" % f, "bytes": size,
            "zone": zones.get(key),
            # A photo with no zone row is the real fault: mockLayout() would
            # fall through to the round-neck tee's measurements.
            "missing": key not in zones,
        })

    # Products drawn without a photo, so the screen accounts for everything
    # rather than silently omitting them.
    no_photo = [names[s] for s in names
                if not any(p["product_id"] == s for p in out)]
    return jsonify(ok=True, photos=out, no_photo=sorted(no_photo),
                   orphan_zones=sorted(k for k in zones
                                       if not any(p["key"] == k for p in out)))


@content_bp.route("/zones/<key>", methods=["POST"])
@require("content")
def set_zone(key):
    from admin_api import log
    d = request.get_json(force=True, silent=True) or {}
    db = get_db()
    cur = db.execute("SELECT * FROM print_zones WHERE mock_key=?", (key,)).fetchone()
    if not cur:
        return jsonify(ok=False, error="No such mockup."), 404
    vals = {}
    # Bounds are sanity, not policy: mockup space is 720px wide, and a print
    # zone measured in metres or millimetres is a typo either way.
    for field, lo, hi in (("cx", 0, 720), ("cy", 0, 900), ("w", 10, 720),
                          ("h", 10, 900), ("cm_w", 1, 200), ("cm_h", 1, 200)):
        v = d.get(field, cur[field])
        try:
            v = float(v)
        except (TypeError, ValueError):
            return jsonify(ok=False, error="%s isn't a number." % field), 400
        if not lo <= v <= hi:
            return jsonify(ok=False,
                           error="%s must be between %g and %g." % (field, lo, hi)), 400
        vals[field] = v
    db.execute("""UPDATE print_zones SET cx=?,cy=?,w=?,h=?,cm_w=?,cm_h=?,
                  updated=CURRENT_TIMESTAMP WHERE mock_key=?""",
               (vals["cx"], vals["cy"], vals["w"], vals["h"],
                vals["cm_w"], vals["cm_h"], key))
    log(db, "zone", 0, "zone_updated", None, None,
        "%s → %g×%g cm" % (key, vals["cm_w"], vals["cm_h"]))
    db.commit()
    return jsonify(ok=True, zone=zones_payload(db).get(key))
